src/test/response_cache_test.cc

```diff
@@ -0,0 +1,598 @@
+  allocated_memory_type_ = *memory_type;
+  allocated_memory_type_id_ = *memory_type_id;
+  allocated_userp_ = nullptr;
+  std::cout << "Done in AllocateDataBuffer" << std::endl;
+  return Status::Success;
+}
+
+Status
+InferenceResponse::AddOutput(
+    const std::string& name, const inference::DataType datatype,
+    const std::vector<int64_t>& shape, InferenceResponse::Output** output)
+{
+  outputs_.emplace_back(name, datatype, shape, allocator_, alloc_userp_);
+
+  // LOG_VERBOSE(1) << "add response output: " << outputs_.back();
+
+  /*if (backend_ != nullptr) {
+    const inference::ModelOutput* output_config;
+    RETURN_IF_ERROR(backend_->GetOutput(name, &output_config));
+    if (output_config->has_reshape()) {
+      const bool has_batch_dim = (backend_->Config().max_batch_size() > 0);
+      outputs_.back().Reshape(has_batch_dim, output_config);
+    }
+  }*/
+
+  if (output != nullptr) {
+    *output = std::addressof(outputs_.back());
+  }
+
+  return Status::Success;
+}
+
+}}  // namespace nvidia::inferenceserver
+
+
+namespace {
+
+// Test Fixture
+class RequestResponseCacheTest : public ::testing::Test {
+  protected:
+    void SetUp() override {}
+    void TearDown() override {}
+  public:
+    ni::InferenceBackend* backend = nullptr;
+    uint64_t model_version = 1;
+};
+
+// Helpers
+void
+check_status(ni::Status status)
+{
+  if (!status.IsOk()) {
+    std::cout << "ERROR: " << status.Message() << std::endl;
+    assert(false);  // TODO
+  }
+}
+
+void
+cache_stats(const ni::RequestResponseCache& cache)
+{
+  std::cout << "Cache entries:" << cache.num_entries() << std::endl;
+  std::cout << "Cache free bytes:" << cache.free_bytes() << std::endl;
+  std::cout << "Cache alloc'd bytes:" << cache.allocated_bytes() << std::endl;
+  std::cout << "Cache total bytes:" << cache.total_bytes() << std::endl;
+}
+
+// Test hashing for consistency on same request
+TEST_F(RequestResponseCacheTest, TestHashing)
+{
+  // Create cache
+  std::cout << "Create cache" << std::endl;
+  uint64_t cache_size = 4 * 1024 * 1024;
+  ni::RequestResponseCache cache(cache_size);
+
+  // Create request
+  std::cout << "Create request" << std::endl;
+  ni::InferenceRequest request0(backend, model_version);
+  ni::InferenceRequest request1(backend, model_version);
+  ni::InferenceRequest request2(backend, model_version);
+
+  // Create inputs
+  std::cout << "Create inputs" << std::endl;
+  inference::DataType dtype = inference::DataType::TYPE_INT32;
+  std::vector<int64_t> shape{1, 4};
+  ni::InferenceRequest::Input* input0 = nullptr;
+  ni::InferenceRequest::Input* input1 = nullptr;
+  ni::InferenceRequest::Input* input2 = nullptr;
+  // Add input to requests
+  std::cout << "Add input to request" << std::endl;
+  request0.AddOriginalInput("input", dtype, shape, &input0);
+  request1.AddOriginalInput("input", dtype, shape, &input1);
+  request2.AddOriginalInput("input", dtype, shape, &input2);
+  assert(input0 != nullptr);
```

Comment:
Should use gtest macro, i.e. `ASSERT_TRUE() / EXPECT_EQ()`. A caveat that I observed is that `XXX_TRUE() / XXX_FALSE()` need to be used when comparing with `nullptr`

Reply:
With gtest macro, you can suffix when output stream semantic to print error message when comparison fails (i.e. `EXPECT_TRUE(false) << "message"`)